src/clang_parser.cpp

```diff
@@ -676,29 +661,68 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
       .Length = btf_cdef.size(),
   });
 
-  ClangParserHandler handler;
-  bool check_additional_types = true;
-  while (check_additional_types && process_btf)
+  if (process_btf)
   {
-    auto incomplete_types = get_incomplete_types(
-        input, input_files, args, bpftrace.btf_set_);
-    size_t types_cnt = bpftrace.btf_set_.size();
-    bpftrace.btf_set_.insert(incomplete_types.cbegin(),
-                             incomplete_types.cend());
-
-    // Update generated header with current BTF types
-    btf_cdef = bpftrace.btf_.c_def(bpftrace.btf_set_);
-    input_files.back() = CXUnsavedFile{
-      .Filename = "/bpftrace/include/__btf_generated_header.h",
-      .Contents = btf_cdef.c_str(),
-      .Length = btf_cdef.size(),
-    };
+    // Resolution of incomplete types must run at least once, maximum should be
+    // the number of levels of nested field accesses for tracepoint args.
+    // The maximum number of iterations can be also controlled by the
+    // BPFTRACE_MAX_TYPE_RES_ITERATIONS env variable (0 is unlimited).
+    unsigned field_lvl = 1;
+    for (auto &probe : *program->probes)
+      if (probe->tp_args_structs_level > (int)field_lvl)
+        field_lvl = probe->tp_args_structs_level;
+
+    unsigned max_iterations = bpftrace.max_type_res_iterations > 0 &&
```

Comment:
Wouldn't std::max(bpftrace.max, field_lvl) be enough here?

Reply:
Of course it would :-) I just over-complicated it, thanks for noticing